### PR TITLE
`<complex>`: fix complex division by zero

### LINKED ISSUE #2728: `<complex>`: Implementation divergence for division by zero
Test case: https://godbolt.org/z/WqzW5b99E

```cpp
#include <complex>
#include <cstdio>
using namespace std;

int main() {
    using C = complex<double>;
    const C t1 = C{1.0} / 0.0;
    const C t2 = C{1.0} / C{0.0};
    printf("    t1: (%g, %g)\n", t1.real(), t1.imag());
    printf("t1 alt: (%g, -%g)\n", t1.real(), -t1.imag());
    printf("    t2: (%g, %g)\n", t2.real(), t2.imag());
}
```

MSVC has always printed (this is not a regression):
```
    t1: (inf, -nan(ind))
t1 alt: (inf, -nan)
    t2: (nan, nan)
```

GCC/libstdc++ and Clang/libc++ print:

```
    t1: (inf, -nan)
t1 alt: (inf, -nan)
    t2: (inf, -nan)
```

This was originally reported as DevCom-10040631 and internal VSO-1539658 / AB#1539658 . The user is wondering whether this is a bug.

I've added "t1 alt" to show that "-nan(ind)" is just how MSVC likes to print negative NaN. Otherwise, the behavior for `C{1.0} / 0.0` is consistent across implementations, and our implementation is just component-wise, so this is unsurprising:

https://github.com/microsoft/STL/blob/1a20fe1133d711a647bbb135d98743f91b7be323/stl/inc/complex#L1465-L1471

For `C{1.0} / C{0.0}`, we go through:

https://github.com/microsoft/STL/blob/1a20fe1133d711a647bbb135d98743f91b7be323/stl/inc/complex#L993-L995

within the `_Div()` function:

https://github.com/microsoft/STL/blob/1a20fe1133d711a647bbb135d98743f91b7be323/stl/inc/complex#L970-L1009

Therefore we're deliberately returning `(nan, nan)` here.

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #2728

tests: https://gcc.godbolt.org/z/66bEr6j4M

### CODE REVIEW

**Comment by fsb4000** [comment]
Do we need copy this?
```
(0,0) / (0,0)     = (-nan,-nan)
(nan,nan) / (0,0) = ( nan, nan)
```
Probably not, I don't know...

File: stl/inc/complex [hunk 1]
```diff
@@ -990,9 +990,28 @@ protected:
...
-            this->_Val[_IM] = this->_Val[_RE];
+        } else if (_Rightimag == 0) { // _Right.real() = 0 && _Right.imag() = 0
+            if (this->_Val[_RE] > 0) {
+                this->_Val[_RE] = _Myctraits::_Infv();
+            } else if (this->_Val[_RE] == 0) {
+                this->_Val[_RE] = -_Myctraits::_Nanv();
+            } else if (this->_Val[_RE] < 0) {
+                this->_Val[_RE] = -_Myctraits::_Infv();
+            } else {
+                // this->_Val[_RE] is NAN
```
Comment (StephanTLavavej) [inline stl/inc/complex:1001]: I suspect (although I'm not sure) that we should test for nan with `_Isnan`, instead of expecting that all of `> 0`, `== 0`, and `< 0` will be false, in order to avoid `/fp:fast` shenanigans. Occurs for the imaginary part below.

This would be consistent with the beginning of `_Div` which tests for nan first.

File: tests/std/tests/Dev10_555491_complex_linker_errors/test.cpp [hunk 2]
```diff
@@ -47,6 +47,87 @@ bool nearly_equal_partwise(const complex<T> actual, const complex<T> expected) {
...
+        assert(isinf(test_minus_one_zero.real()) && test_minus_one_zero.real() < 0);
+        assert(isnan(test_minus_one_zero.imag()));
+    }
+    {
+        const complex<double> test_minus_one_minus_one = complex<double>{-1, -1} / zero;
+        assert(isinf(test_minus_one_minus_one.real()) && test_minus_one_minus_one.real() < 0);
+        assert(isinf(test_minus_one_minus_one.imag()) && test_minus_one_minus_one.imag() < 0);
+    }
+    {
+        const complex<double> test_inf_inf = complex<double>{INFINITY, INFINITY} / zero;
```
Comment (StephanTLavavej) [inline tests/std/tests/Dev10_555491_complex_linker_errors/test.cpp:99]: We generally avoid the `INFINITY` and `NAN` macros in test code (the CRT defines them in a somewhat unusual way and this has changed over time; `numeric_limits` uses builtins with predictable behavior); below this test uses `constexpr double inf = numeric_limits<double>::infinity();` and we could do the same here.

File: tests/std/tests/Dev10_555491_complex_linker_errors/test.cpp [hunk 3]
```diff
@@ -47,6 +47,87 @@ bool nearly_equal_partwise(const complex<T> actual, const complex<T> expected) {
     return nearly_equal(actual.real(), expected.real()) && nearly_equal(actual.imag(), expected.imag());
 }
 
+// Also test GH-2728 <complex>: Implementation divergence for division by zero
+void test_gh_2728() {
+    const complex<double> zero{0, 0};
+    {
+        const complex<double> test_one_one = complex<double>{1, 1} / zero;
+        assert(isinf(test_one_one.real()) && test_one_one.real() > 0);
```
Comment (StephanTLavavej) [inline tests/std/tests/Dev10_555491_complex_linker_errors/test.cpp:55]: It's not a big deal here, but as a general rule, it's better to separately assert conditions with `assert(is_cute); assert(is_fluffy);` instead of `assert(is_cute && is_fluffy);`. That way, if the code ever encounters a cute Sphynx kitten, or an ugly fluffy kitten, then the assertion line will indicate exactly what went wrong.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for fixing this high-pri DevCom bug! :smile_cat: I went ahead and pushed fairly minor changes (the most notable one being avoiding any possible `/fp:fast` impact).

File: stl/inc/complex [hunk 4]
```diff
@@ -990,9 +990,26 @@ protected:
...
+
+            if (_Myctraits::_Isnan(this->_Val[_IM])) {
+                this->_Val[_IM] = _Myctraits::_Nanv();
+            } else if (this->_Val[_IM] > 0) {
+                this->_Val[_IM] = _Myctraits::_Infv();
+            } else if (this->_Val[_IM] == 0) {
+                this->_Val[_IM] = -_Myctraits::_Nanv();
+            } else { // this->_Val[_IM] < 0
+                this->_Val[_IM] = -_Myctraits::_Infv();
+            }
```
Comment (statementreply) [inline stl/inc/complex:1012]:
```suggestion
            this->_Val[_RE] /= _Rightreal;
            this->_Val[_IM] /= _Rightreal;
```
There are two different zeroes in (real) IEEE-754 binary floating point, +0 and -0. Under the default mode, _x_/(-_y_) = -(_x_/_y_) is preserved even when _y_ is zero and _x_ is nonzero: +1/+0 = -1/-0 = +∞, -1/+0 = +1/-0 = -∞.

Each of the real and imaginary parts of a `complex` zero can be either +0 or -0. As a result, there are four different zeros in `complex`, (+0,+0), (+0,-0), (-0,+0) and (-0,-0). It would be nice that complex division by zero also preserves _z_/(-_w_) = -(_z_/_w_): _z_/(-0,+0) = -(_z_/(+0,-0)), _z_/(-0,-0) = -(_z_/(+0,+0)). Dividing by `_Rightreal` is one of the possible ways to achieve this.

Additionally, division by zero raises `FE_DIVBYZERO` or `FE_INVALID` floating-point exception, but `_Ctraits::_Infv()` or `_Ctraits::_Nanv()` doesn't. (https://godbolt.org/z/xzac4G4j8)

ISO C [G.5.1]/5
> If both operands of the `*` operator are complex or if the second operand of the `/` operator is complex, the operator raises floating-point exceptions if appropriate for the calculation of the parts of the result, and may raise spurious floating-point exceptions.

**Review by StephanTLavavej — APPROVED** [review]

**Review by strega-nil-ms — APPROVED** [review]
LGTM!

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed.

**Comment by StephanTLavavej** [comment]
Thanks for fixing this high priority user-reported bug! Who knew that division by zero was so important! :zero: :joy_cat: :tada:

**Comment by fsb4000** [comment]
Yeah, I have failed to change 2 lines of the product code :)
Thanks @statementreply for suggestion the correct fix 🎉
